server/service/src/programs/patient/insert_name_patient.rs

```diff
@@ -0,0 +1,73 @@
+use repository::{
+    EqualFilter, NameRow, NameRowRepository, NameType, Patient, PatientFilter, RepositoryError,
+    StorageConnection, TransactionError,
+};
+
+use crate::service_provider::{ServiceContext, ServiceProvider};
+
+#[derive(PartialEq, Debug)]
+pub enum InsertNamePatientError {
+    PatientExists,
+    NotAPatient,
+    InternalError(String),
+    DatabaseError(RepositoryError),
+}
+
+fn validate_patient_does_not_exist(
+    con: &StorageConnection,
+    input: &NameRow,
+) -> Result<bool, RepositoryError> {
+    let existing = NameRowRepository::new(con).find_one_by_id(&input.id)?;
+    Ok(existing.is_none())
+}
+
+fn validate(con: &StorageConnection, input: &NameRow) -> Result<(), InsertNamePatientError> {
+    if input.r#type != NameType::Patient {
+        return Err(InsertNamePatientError::NotAPatient);
+    }
+
+    if !validate_patient_does_not_exist(con, input)? {
+        return Err(InsertNamePatientError::PatientExists);
+    }
+    Ok(())
+}
+
+pub(crate) fn insert_name_patient(
+    ctx: &ServiceContext,
+    service_provider: &ServiceProvider,
+    input: &NameRow,
+) -> Result<Patient, InsertNamePatientError> {
+    let patient = ctx
+        .connection
+        .transaction_sync(|con| {
+            validate(con, &input)?;
+
+            let name_repo = NameRowRepository::new(con);
+            name_repo.upsert_one(input)?;
+
+            let patient = service_provider
+                .patient_service
+                .get_patients(
+                    ctx,
+                    None,
+                    Some(PatientFilter::new().id(EqualFilter::equal_to(&input.id))),
+                    None,
+                    None,
+                )
+                .map_err(|err| InsertNamePatientError::DatabaseError(err))?
+                .rows
+                .pop()
+                .ok_or(InsertNamePatientError::InternalError(
```

Comment:
Disagree :) this error wouldn't be useful and would even be misleading to the caller since this error should never happen. If we just inserted a patient and then can't read it in the same tx (and there wasn't another error) there is an internal bug. This internal error shouldn't leak out other than saying the operation failed.

The error is implementation specific and shouldn't be part of the service "API/interface". For example, you could construct the returned patient just from the inserted input without reading it from the DB and then this error would go away; service users shouldn't be affected by that.

Reply:
Hmm, I think in that case it's ok to unwrap() ? Since error would never happen or leak out anyways ?'

I get your point about something impossible being expose in service API and it's convincing.

Reply:
Yes agree unwrap would be ok here. On the other hand I also like that we minimize the use of unwrap(), e.g. when are we really 100% that we can unwrap()?

Can't remember why but I recently had an endpoint that paniced for some reasons and then the error on the frontend was a bit confusing because I haven't seen it for such a long time :-)

Reply:
My 2cents on this.
(Not intending to re-spark this conversation)

I've been experimenting with adding a `GenericError` enum option and I've found it feels like it's faster to develop, and I spend less time mapping new enum variants though lots of files etc. Especially when you want to just try out a bit of code.
I think it is a bit dangerous though, as it tends to make my/us lazy with thinking about the errors...

Here's an interesting example...
https://github.com/openmsupply/health-supply-hub/blob/0e92eaad18e51bacc7a5d3f8448625de85debc6d/backend/service/src/organisation_group/create.rs#L68

These errors aren't `InternalError` and should map to `BadUserInput` in graphql.

My feeling is that things that make the development process easier and help avoid issues like when exactly we should `unwrap` or `panic` is probably on balance a good thing.

Where we really need concrete types are when we want calling code to match on the types and handle the error.
To me this is a good `what we should do `example`

```
#[derive(Debug)]
pub enum TemporaryErrorType {
    TimedOut(String),
    ConnectionError(String),
    InternalServerError(String),
    Flood(String),
    Other(String),
}

#[derive(Debug)]
pub enum TelegramError {
    Fatal(String),
    Temporary(TemporaryErrorType),
}
```
https://github.com/openmsupply/notify/blob/8d5a33817e706a193b4a23c6447cc6a14cc69609/backend/telegram/src/client/mod.rs#L24C1-L28C2

This lets calling code handle `Fatal` errors differently to `Permanent`, which is where I think the matching in rust really shines.

I like Clemens `InternalError` naming rather than `GenericError` which I've been using.

Here's an example graphql mapping function with lots of error variants vs one using generic error

```
fn map_error(error: ModifyUserAccountError) -> Result<CreateUserAccountResponse> {
    use StandardGraphqlError::*;
    let formatted_error = format!("{:#?}", error);

    let graphql_error = match error {
        // Standard Graphql Errors
        ModifyUserAccountError::UserAccountAlreadyExists => BadUserInput(formatted_error),
        ModifyUserAccountError::EmailAddressAlreadyExists => BadUserInput(formatted_error),
        ModifyUserAccountError::UserAccountDoesNotExist => BadUserInput(formatted_error),
        ModifyUserAccountError::InvalidPassword => BadUserInput(
            "Please ensure your password meets the minimum complexity standards".to_string(),
        ),
        ModifyUserAccountError::InvalidUsername => BadUserInput(
            "Invalid Username, username can only include english letters and numbers".to_string(),
        ),
        ModifyUserAccountError::PermissionsMissing => {
            BadUserInput("User must have a role or permission assigned".to_string())
        }
        ModifyUserAccountError::DatabaseError(_) => InternalError(formatted_error),
        ModifyUserAccountError::ModifiedRecordNotFound => InternalError(formatted_error),
        ModifyUserAccountError::PasswordHashError(_) => InternalError(formatted_error),
        ModifyUserAccountError::UnableToSendEmail(_) => InternalError(formatted_error),
        ModifyUserAccountError::InvalidToken => BadUserInput(formatted_error),
        ModifyUserAccountError::TokenExpired => {
            BadUserInput("Token has expired, please request a new invite".to_string())
        }
    };

    Err(graphql_error.extend())
}
```
https://github.com/openmsupply/health-supply-hub/blob/0e92eaad18e51bacc7a5d3f8448625de85debc6d/backend/graphql/user_account/src/mutations/create.rs#L81
```
fn map_error(error: ServiceError) -> Result<InsertOrganisationGroupResponse> {
    use StandardGraphqlError::*;
    let formatted_error = format!("{:#?}", error);

    let graphql_error = match error {
        // Standard Graphql Errors
        ServiceError::OrganisationGroupAlreadyExists => BadUserInput(formatted_error),
        ServiceError::DatabaseError(_) => InternalError(formatted_error),
        ServiceError::GenericError(_) => BadUserInput(formatted_error),
        ServiceError::ModifiedRecordNotFound => BadUserInput(formatted_error),
        ServiceError::OrganisationGroupDoesNotExist => BadUserInput(formatted_error),
    };

    Err(graphql_error.extend())
}
```
https://github.com/openmsupply/health-supply-hub/blob/0e92eaad18e51bacc7a5d3f8448625de85debc6d/backend/graphql/organisation_group/src/mutations/insert.rs#L74C1-L88C2

Personally I think the service layer is probably best placed to define the string the user will see in the UI for example in user account example we have `Invalid Username, username can only include english letters and numbers` however in the service layer we've actually modified the regex, so email addresses are now valid usernames. We're more likely to capture this and update the error string in the service layer I think.

My proposal is something like this.

It's ok to use `InternalError(String)` and `BadUserInput(String)` error variants in service layer if a developer finds it useful.
Where a calling function might need to handle a specific error differently a specific variant should be used to allow matching.
E.g.
```
pub enum TelegramError {
    Fatal(String),
    Temporary(TemporaryErrorType),
}
```

I think/hope having a distinction between `InternalError` and `BadUserInput` will help prompt the develop to think about how the error will be used and what it might mean vs `GenericError` which might be used in different ways.

Translating these strings is obviously an issue, here's a few thoughts.
1. If we are returning an `InternalError` I think this is ok to be in english (assuming that something has gone seriously wrong to cause and internal error, so the end user is likely to need to get support involved anyway, or it's needed to help support development)
2. For `BadUserInput` what if we have a pattern using a BadUserInput Struct something like this?
```
pub struct UserInputError {
pub message: String,
pub translation_key: String
}
```
If we propagate this all the way through to the frontend, we could have a component that translates UserError?
There's potentially a better approach than this, but if feels workable and good value for effort?

Reply:
Pretty sure we discussed this or at least very similar as well :-) We setted to be explicite about what errors can happen in the service layer but have the compromise to have standard errors in the graphql layer (because the mapping there became very tedious).

As you described, mapping in graphql layer could be shorter but its also nice to be forced to think about what a service error means for to the API user, e.g. if the error needs to be handled specially (but I guess this could already be done in the service layer as you showed).

Yes think it would be cool to have a unique constant `key` or `code` field in the standard error details which could be used for translation if needed. But then you have to come up with a unique schema for every error or somehow derive it automatically, e.g. from the enum name?